### PR TITLE
Fix reply mention oddities

### LINKED ISSUE #833: Editing a reply message which didn't mention makes it mention
# Summary
If you send a reply message that does not mention, then edit it, it will make it mention. This is not normal behaviour.

# Details
No idea why that is, I reckon @VelvetThePanda and @ProfDoof will have some input as they are the ones who originally implemented the feature in D#+

# Steps to reproduce
```cs
[Command]
public async Task TestReplyBork(CommandContext ctx)
{
    var msg = await ctx.RespondAsync("nope"); // sends a non-mentioning reply by default
    await msg.ModifyAsync(x => x.Content = "yep"); // now it mentions
}
```

# Notes
This has to be resolved before 4.0 release.

### PR DESCRIPTION
**Pull request by VelvetThePanda**
# Summary
Provides a proper fix to #1044, which would've introduced more bugs regarding mentions.

# Details
This fixes an issue where editing a message would absolutely wreck mentions, which is [at least partly Discord's fault](https://github.com/discord/discord-api-docs/issues/3378)

This also closes #833.

# Changes proposed
Fix the use of `DiscordMentions`, which was the *actual* cause of the reply bug.
Add `GetMentions` method to DiscordMessage, which resolves what entities have actually been mentioned.
Remove parsing users by regex as per [this comment](https://github.com/DSharpPlus/DSharpPlus/pull/1044#issuecomment-903911763). It *could* be a breaking change.
Added a `_rawMentionedRoles` field to aid in figuring out what roles are actually mentioned, because `_mentionedRoles` contains API-provided mentions + regex'd mentions.

# Notes
Closes #1044, Closes #833 (For GitHub)

I've tested this manually for the past hour figuring out how to get it to work, so I'm gonna hope I didn't break it.

### CODE REVIEW

**Comment by VelvetThePanda** [comment]
Okay I'm gonna fix an issue with role mentions

**Comment by VelvetThePanda** [comment]
Ready for review/merge.

File: DSharpPlus/Entities/Channel/Message/DiscordMessageBuilder.cs [hunk 1]
```diff
@@ -397,8 +397,8 @@ internal void Validate(bool isModify = false)
 
             if (isModify)
             {
-                if (this.ReplyId.HasValue)
-                    throw new ArgumentException("You cannot change the ReplyID when modifying a message");
+                if (this.ReplyId.HasValue) ;
+                //throw new ArgumentException("You cannot change the ReplyID when modifying a message");
```
Comment (Lulalaby) [inline DSharpPlus/Entities/Channel/Message/DiscordMessageBuilder.cs:401]: Why exactly did you removed this

Comment (VelvetThePanda) [reply DSharpPlus/Entities/Channel/Message/DiscordMessageBuilder.cs:401]: See #1059

Comment (Lulalaby) [reply DSharpPlus/Entities/Channel/Message/DiscordMessageBuilder.cs:401]: ah ok

**Review by Lulalaby — APPROVED** [review]

File: DSharpPlus/Entities/Channel/Message/DiscordMessage.cs [hunk 2]
```diff
@@ -391,7 +398,21 @@ internal DiscordMessageReference InternalBuildMessageReference()
             return reference;
         }
 
+        private IMention[] GetMentions()
+        {
+            var mentions = new List<IMention>();
+
+            if (this.ReferencedMessage != null && this._mentionedUsers.Contains(this.ReferencedMessage.Author))
+               mentions.Add(new RepliedUserMention()); // Return null to allow all mentions
```
Comment (Neuheit) [inline DSharpPlus/Entities/Channel/Message/DiscordMessage.cs:406]: Do you intend for this to be the only thing that is added to `mentions` if it contains the author?

Comment (VelvetThePanda) [reply DSharpPlus/Entities/Channel/Message/DiscordMessage.cs:406]: No I don't think so? It worked in testing, but I'll look again

Comment (Neuheit) [reply DSharpPlus/Entities/Channel/Message/DiscordMessage.cs:406]: Because the mentioneduser check will run right below it and add the user still.

**Comment by VelvetThePanda** [comment]
@Neuheit apparently what I wrote was what I intended; I don't know how I did it, but I managed to preserve all mentions. Kind of black magic if you ask me now.

I don't think it's entirely desirable, but if you explicitly state the reply author in allowed mentions but set the reply to false, the previous behavior is still exhibited, but I find that situation daft. 

Anyway, for the most part it's fixed, as shown here:

Image

cc @Emzi0767

**Review by Emzi0767 — APPROVED** [review]

**Comment by Emzi0767** [comment]
There are conflicts afoot.

**Pull request merged**